jupyterlab/extensions/pypi.py

```diff
@@ -0,0 +1,405 @@
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+"""Extension manager using pip as package manager and PyPi.org as packages source."""
+
+import asyncio
+import io
+import json
+import re
+import sys
+import xmlrpc.client
+from datetime import datetime, timedelta
+from functools import partial
+from itertools import groupby
+from pathlib import Path
+from subprocess import CalledProcessError, run
+from tarfile import TarFile
+from typing import Any, Callable, Dict, List, Optional, Tuple
+from zipfile import ZipFile
+
+import tornado
+from async_lru import alru_cache
+from traitlets import CFloat, Unicode, config
+
+from jupyterlab.extensions.manager import (
+    ActionResult,
+    ExtensionManager,
+    ExtensionManagerMetadata,
+    ExtensionPackage,
+)
+
+
+@alru_cache(maxsize=500)
```

Comment:
1. While there are 167 extensions with `FRAMEWORK :: JUPYTER :: JUPYTERLAB :: EXTENSIONS :: PREBUILT` modifier, there are 1565 packages matching `jupyterlab` on NPM. After release of the new extension manager we can expect developers to port more extensions to prebuilt. Raising maxsize to 1500 should not hurt (we can revisit later):

```suggestion
@alru_cache(maxsize=1500)
```

2. Can we make it configurable by keeping the standalone function `_fetch_package_metadata` in global scope without decorator and add a decorator dynamically in `__init__` of `PyPiExtensionManager` based on a value from configurable tratitlet? Something like:

```python
class PyPiExtensionManager(ExtensionManager):
   metadata_cache = Integer(1500, config=True, help="The cache size for package metadata")
   def __init__(self):
       super.__init()
       this._fetch_package_metadata = alru_cache(maxsize=self.metadata_cache)(_fetch_package_metadata)